### ISSUE TITLE
Why Root Widget gets build twice?

### ISSUE THREAD

**Issue by rajeshpatil95** [flutter/flutter#80225]
https://stackoverflow.com/questions/55515371/why-root-widget-gets-build-twice

Inside build method just displaying container without state being toggled....??

Flutter Doctor:
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 2.0.4, on Linux, locale en_IN)
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
[✓] Chrome - develop for the web
[✓] Android Studio (version 4.0)
[✓] VS Code (version 1.54.3)
[✓] Connected device (2 available)

• No issues found!

**Comment by iapicca** [comment]
@rajeshpatil95 
using the code sample provided in the [SO question ](https://stackoverflow.com/questions/55515371/why-root-widget-gets-build-twice) linked [above](https://github.com/flutter/flutter/issues/80225#issue-855364465)

<details>
<summary>sample</summary>

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {

  @override
  Widget build(BuildContext context) {
    print("ROOT WIDGET ");
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Container(decoration: BoxDecoration(color: Colors.blue),),
    );
  }
}
```

</details>

the issue does not occur testing on physical android device with the latest master 

<details>
<summary>logs</summary>

```bash
[   +3 ms] Installing build/app/outputs/flutter-apk/app.apk...
[        ] executing: /home/francesco/Android/Sdk/platform-tools/adb -s 965AY0WP5C install -t -r /home/francesco/projects/issue/build/app/outputs/flutter-apk/app.apk
[+3385 ms] Performing Streamed Install
                    Success
[   +1 ms] Installing build/app/outputs/flutter-apk/app.apk... (completed in 3.4s)
[   +3 ms] executing: /home/francesco/Android/Sdk/platform-tools/adb -s 965AY0WP5C shell echo -n 55b6f046cdd94238a5cb6a0696a4805d8c9cb5b3 > /data/local/tmp/sky.com.example.issue.sha1
[  +45 ms] executing: /home/francesco/Android/Sdk/platform-tools/adb -s 965AY0WP5C shell -x logcat -v time -t 1
[  +55 ms] --------- beginning of main
           04-11 23:33:56.718 I/Finsky  (24069): [2] acde.g(8): Invalidating cached PackageState for com.example.issue
[  +15 ms] executing: /home/francesco/Android/Sdk/platform-tools/adb -s 965AY0WP5C shell am start -a android.intent.action.RUN -f 0x20000000 --ez enable-background-compilation true --ez
enable-dart-profiling true --ez enable-checked-mode true --ez verify-entry-points true com.example.issue/com.example.issue.MainActivity
[ +139 ms] Starting: Intent { act=android.intent.action.RUN flg=0x20000000 cmp=com.example.issue/.MainActivity (has extras) }
[   +1 ms] Waiting for observatory port to be available...
[+1355 ms] Observatory URL on device: http://127.0.0.1:39081/yg_RTLNRvl8=/
[   +4 ms] executing: /home/francesco/Android/Sdk/platform-tools/adb -s 965AY0WP5C forward tcp:0 tcp:39081
[  +15 ms] 37277
[        ] Forwarded host port 37277 to device port 39081 for Observatory
[  +10 ms] Caching compiled dill
[ +176 ms] Connecting to service protocol: http://127.0.0.1:37277/yg_RTLNRvl8=/
[ +569 ms] Launching a Dart Developer Service (DDS) instance at http://127.0.0.1:0, connecting to VM service at http://127.0.0.1:37277/yg_RTLNRvl8=/.
[ +129 ms] I/flutter (25275): ROOT WIDGET 
[ +161 ms] DDS is listening at http://127.0.0.1:40161/gOJ5G7KaNvc=/.
[  +62 ms] Successfully connected to service protocol: http://127.0.0.1:37277/yg_RTLNRvl8=/
[  +91 ms] DevFS: Creating new filesystem on the device (null)
[  +60 ms] DevFS: Created new filesystem on the device (file:///data/user/0/com.example.issue/code_cache/issueNFHBVZ/issue/)
[   +2 ms] Updating assets
[  +91 ms] Syncing files to device Pixel 3a...
[   +1 ms] <- reset
[        ] Compiling dart to kernel with 0 updated files
[   +2 ms] <- recompile package:issue/main.dart bf78d180-1180-4194-8803-d8c5da5b09aa
[        ] <- bf78d180-1180-4194-8803-d8c5da5b09aa
[  +90 ms] Updating files.
[        ] DevFS: Sync finished
[   +1 ms] Syncing files to device Pixel 3a... (completed in 94ms)
[        ] Synced 0.0MB.
[        ] <- accept
[  +11 ms] Connected to _flutterView/0x77f3fc17b0.
[   +2 ms] Flutter run key commands.
[   +1 ms] r Hot reload. 🔥🔥🔥
[        ] R Hot restart.
[        ] h Repeat this help message.
[        ] d Detach (terminate "flutter run" but leave application running).
[        ] c Clear the screen
[        ] q Quit (terminate the application on the device).
[        ] 💪 Running with sound null safety 💪
[        ] An Observatory debugger and profiler on Pixel 3a is available at: http://127.0.0.1:40161/gOJ5G7KaNvc=/
[+1083 ms] Activating Dart DevTools...
[+2542 ms] Activating Dart DevTools... (completed in 2,541ms)
[ +662 ms] The Flutter DevTools debugger and profiler on Pixel 3a is available at: http://127.0.0.1:9102?uri=http%3A%2F%2F127.0.0.1%3A40161%2FgOJ5G7KaNvc%3D%2F
[+23664 ms] DevFS: Deleting filesystem on the device (file:///data/user/0/com.example.issue/code_cache/issueNFHBVZ/issue/)
[  +36 ms] DevFS: Deleted filesystem on the device (file:///data/user/0/com.example.issue/code_cache/issueNFHBVZ/issue/)
[ +346 ms] Service protocol connection closed.
[        ] Application finished.
[   +3 ms] executing: /home/francesco/Android/Sdk/platform-tools/adb -s 965AY0WP5C forward --list
[   +5 ms] Exit code 0 from: /home/francesco/Android/Sdk/platform-tools/adb -s 965AY0WP5C forward --list
[        ] 965AY0WP5C tcp:37277 tcp:39081
[        ] executing: /home/francesco/Android/Sdk/platform-tools/adb -s 965AY0WP5C forward --remove tcp:37277
[   +9 ms] executing: /home/francesco/Android/Sdk/platform-tools/adb -s 965AY0WP5C forward --list
[   +6 ms] Exit code 0 from: /home/francesco/Android/Sdk/platform-tools/adb -s 965AY0WP5C forward --list
[   +2 ms] "flutter run" took 98,536ms.
[ +259 ms] ensureAnalyticsSent: 256ms
[   +6 ms] Running shutdown hooks
[   +1 ms] Shutdown hooks complete
[   +5 ms] exiting with code 0

```

</details>

<details>
<summary>doctor</summary>

```console
[✓] Flutter (Channel master, 2.1.0-13.0.pre.574, on Linux, locale en_US.UTF-8)
    • Flutter version 2.1.0-13.0.pre.574 at /home/francesco/snap/flutter/common/flutter
    • Framework revision 02efffc134 (2 days ago), 2021-04-10 03:49:01 -0400
    • Engine revision 8863afff16
    • Dart version 2.13.0 (build 2.13.0-222.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /home/francesco/Android/Sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_SDK_ROOT = /home/francesco/Android/Sdk
    • Java binary at: /usr/lib/jvm/java-1.8.0-openjdk-amd64/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_282-8u282-b08-0ubuntu1~20.04-b08)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = /snap/bin/chromium

[✓] Linux toolchain - develop for Linux desktop
    • clang version 10.0.0-4ubuntu1
    • cmake version 3.16.3
    • ninja version 1.10.0
    • pkg-config version 0.29.1

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/linux#android-setup for detailed instructions).

[✓] Connected device (3 available)
    • Pixel 3a (mobile) • 965AY0WP5C • android-arm64  • Android 11 (API 30)
    • Linux (desktop)   • linux      • linux-x64      • Linux
    • Chrome (web)      • chrome     • web-javascript • Chromium 89.0.4389.114 snap

! Doctor found issues in 1 category.

```

</details>

**Comment by darshankawar** [comment]
@rajeshpatil95 
As mentioned in above comment, the root widget is building only once as expected:

```
Running "flutter pub get" in macbug...
Launching lib/main.dart on iPhone 12 Pro Max in debug mode...
Running Xcode build...
Xcode build done.                                           23.8s
Debug service listening on ws://127.0.0.1:53517/pMKNhz5YTrA=/ws
Syncing files to device iPhone 12 Pro Max...
flutter: ROOT WIDGET

```

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 2.0.4, on Mac OS X 10.15.4 19E2269 darwin-x64,
    locale en-GB)
    • Flutter version 2.0.4 at /Users/dhs/documents/fluttersdk/flutter
    • Framework revision b1395592de (7 days ago), 2021-04-01 14:25:01 -0700
    • Engine revision 2dce47073a
    • Dart version 2.12.2

[!] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    ! CocoaPods 1.9.3 out of date (1.10.0 is recommended).
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin
        code that responds to your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To upgrade see
      https://guides.cocoapods.org/using/getting-started.html#installation for
      instructions.

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome
[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)    

[✓] VS Code (version 1.55.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.20.0

[✓] Connected device (3 available)
    • SM A260G (mobile)            • 5200763ebcfa861f • android-arm    • Android
      8.1.0 (API 27)
    • Darshan's iphone (mobile) • 21150b119064aecc249dfcfe05e259197461ce23 • ios
      • iOS 14.4.1
    • macOS (desktop)           • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)              • chrome                                   •
      web-javascript • Google Chrome 89.0.4389.114

! Doctor found issues in 1 category.

```

</details>

**Comment by pedromassangocode** [comment]
Without additional information, we are unfortunately not sure how to resolve this issue. 
We are therefore reluctantly going to close this bug for now. 
Please don't hesitate to comment on the bug if you have any more information for us; we will reopen it right away!  
Thanks for your contribution.

Could everyone who still has this problem please file a new issue with the exact description of what happens, logs, and the output of `flutter doctor -v`.
All system setups can be slightly different, so it's always better to open new issues and reference related issues.
